Source/Particles/Collision/BinaryCollision.cpp

```diff
@@ -191,13 +196,12 @@ void PairWiseCoulombCollision::doCoulombCollisionsWithinTile
                 auto dV = MathConst::pi*(2.0_rt*ri+1.0_rt)*dr*dr*dz;
 #endif
                 // Call the function in order to perform collisions
-                ElasticCollisionPerez(
+                (*m_binary_collision_functor)(
```

Comment:
Do I understand correctly that this uses a virtual function within a GPU kernel here?

If yes, I think that it would probably better (for performance) to instead use a template-based solution. We do something similar e.g. for ionization, where we pass ionization functors to the function `filterCopyTransformParticles`, which is templated on the functor type:
https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/MultiParticleContainer.cpp#L864

Reply:
Note that it is possible that this particle part of the code is the one that is causing the CI failure for NVCC and HIP:
https://github.com/ECP-WarpX/WarpX/pull/2057/checks?check_run_id=2997511111

Reply:
Absolutely, this was not working with GPUs. Now we use a templated `BinaryCollision` class as you and Axel suggested.